### DISCUSSION TITLE
How to use async values when configuring providers?

### DISCUSSION THREAD

**Discussion by patatoa** [nextauthjs/next-auth#2394] [Help] [answered]
When I set up my providers, I need to pull some secrets from Google Cloud's Secret Manager. This is an async operation and it is giving me quite a headache trying to solve.
``` js
export default NextAuth({
  providers: [
    Providers.IdentityServer4({
        id: "identity-server", //...
        clientSecret: await GetSecret(), //...
    })
  ],
```
Obviously that won't work. 

I also tried something like 
``` js
const secret = (async () => await GetSecret())();
export default NextAuth({
  providers: [
    Providers.IdentityServer4({
        id: "identity-server", //...
        clientSecret: secret, //...
    })
  ],
```
but that *still* isn't working, though it compiles and is valid code.

I cannot store this in a .env or app.yaml file, because I'd need to be able to check these files in while still keeping secrets secure.
Any idea how I can use this async value when setting up the providers?

Thank you for your help!

**Comment by balazsorban44** [comment, marked as answer]
This should work:

```js
// /pages/api/auth/[...nextauth].js
export default async function handler(req, res) {
  NextAuth(req, res, {
    providers: [
      Providers.IdentityServer4({
          id: "identity-server", //...
          clientSecret: await GetSecret(),
      })
    ]
  })
}
```

**Reply by balazsorban44** [reply]
@ndom91 Maybe we could mention this alternative way of initializing in the docs.

This has been the default before (see #867), but it is not needed in most cases, and so it should be reserved when users have special cases like the above.

**Reply by patatoa** [reply]
This solved it! Thank you!

**Reply by ndom91** [reply]
Whipped up a quick PR with this note. See: https://github.com/nextauthjs/next-auth/pull/2443
